Search in Launchpad by PR URL. GitHub (#3543)

sergeibbb opened the pull request:
# Description

When I copy and paste a PR URL directly into the input box on the main step, it should search for that PR

# Checklist

<!-- Please check off the following -->

- [x] I have followed the guidelines in the Contributing document
- [x] My changes follow the coding style of this project
- [x] My changes build without any errors or warnings
- [x] My changes have been formatted and linted
- [x] My changes include any required corresponding changes to the documentation (including CHANGELOG.md and README.md)
- [x] My changes have been rebased and squashed to the minimal number of relevant commits (~~typically 1~~ , **don't squash my lovely commits,** this PR is too big too be squashed).
- [x] My changes have a descriptive commit message with a short title, including a `Fixes $XXX -` or `Closes #XXX -` prefix to auto-close the issue that your PR addresses

sergiolms commented on src/plus/integrations/providers/github.ts:
```diff
@@ -317,3 +318,52 @@ export class GitHubEnterpriseIntegration extends GitHubIntegrationBase<SelfHoste
+	let ownerAndRepo: string | undefined = undefined;
+	let prNumber: string | undefined = undefined;
+
+	let match = search.match(/github\.com\/([^/]+\/[^/]+)\/pull\/(\d+)/); // with github.com domain
+	if (match != null) {
+		ownerAndRepo = match[1];
+		prNumber = match[2];
+	}
+
+	if (prNumber == null) {
+		match = search.match(/([^/]+\/[^/]+)\/pull\/(\d+)/); // without domain
+		if (match != null) {
+			ownerAndRepo = match[1];
+			prNumber = match[2];
+		}
+	}
```
Is there a reason for checking with the `github.com` domain first? The matching groups don't seem to be different, you could directly run the search without the domain, right? If the URL has github domain in it, it will work anyway with just the second query.
![Captura de pantalla 2024-10-21 a las 10 51 46](https://github.com/user-attachments/assets/d04af815-586c-4ebc-b0f5-70e937d659c3)

sergeibbb replied:
@sergiolms 
correct. fixed.

sergiolms reviewed (commented):
There are two odd things I've found while testing this. I haven't looked at the whole code yet:

1) I tried to paste this link: `https://github.com/gitkraken/vscode-gitlens/pull/3684/files#diff-e4b3bb605be44151fa3b67bc6a53b6178113c32d7f5ae627fd1b8fb4ba3e1a61R331-R343` but it didn't work until I removed the #diff... part:

https://github.com/user-attachments/assets/dc4847d3-fda2-4cb2-8a92-f26fd9d048dd

However it should be getting the correct parts 🤔 
![Captura de pantalla 2024-10-21 a las 11 15 56](https://github.com/user-attachments/assets/519dbbc8-e182-42b6-afb2-53612964c689)


2) After opening a URL in the browser, some items disappear but they still count, creating some empty items you cannot interact with:



https://github.com/user-attachments/assets/36cc57e0-6b4a-40ca-87ca-852485bea085

sergeibbb commented:
@sergiolms 

> I tried to paste this link: https://github.com/gitkraken/vscode-gitlens/pull/3684/files#diff-e4b3bb605be44151fa3b67bc6a53b6178113c32d7f5ae627fd1b8fb4ba3e1a61R331-R343 but it didn't work until I removed the #diff... part:

Thank you. Fixed

sergeibbb commented:
@sergiolms 

> After opening a URL in the browser, some items disappear but they still count, creating some empty items you cannot interact with:

Good catch! I should not search on the next step.

sergeibbb commented:
@sergiolms

> After opening a URL in the browser, some items disappear but they still count, creating some empty items you cannot interact with:

It looks like existing behavior. Got to an item and put a query that doesn't match any command and these two item appear.


https://github.com/user-attachments/assets/1983a2e7-21e4-4453-aab8-a392873253fa

axosoft-ramint commented on src/system/vscode/asyncDebouncer.ts:
```diff
@@ -0,0 +1,134 @@
+import type { Deferred } from '../promise';
+import { defer } from '../promise';
+
+export interface AsyncTask<T> {
+	(cancelationToken: CancellationToken): T | Promise<T>;
+}
+
+/**
+ * This is similar to `src/system/function.ts: debounce` but it's for async tasks.
+ * The old `debounce` function does not awaits for promises, so it's not suitable for async tasks.
+ *
+ * Here the debouncer returns a promise that awaits task for completion.
+ * Also we can let tasks know if they are cancelled by passing a cancellation token.
+ *
+ * Despite being able to accept synchronous tasks, we always return a promise here. It's implemeted this way for simplicity.
+ */
+export function createAsyncDebouncer<T>(delay: number): Disposable & Deferrable<(task: AsyncTask<T>) => Promise<T>> {
```
Would it work to move this to the same `src/system/function` file? It appears to me like this is another utility for working with functions.

sergeibbb replied:
@axosoft-ramint 

I wanted to do this. But `src/system/function.ts` is supposed to be used both in vscode and in web-view. While `asyncDebouncer` imports `CancellationTokenSource` from `vscode`, so placing it to `function.ts` makes the project non-compilable.

Therefore I've put it under `src/system/vscode`.

axosoft-ramint replied:
I see. Thanks for explaining 👍🏼

axosoft-ramint commented on src/system/function.ts:
```diff
@@ -98,13 +97,6 @@ export function debounce<T extends (...args: any[]) => ReturnType<T>>(
 		lastThis = this;
 		lastCallTime = time;
 
-		if (isInvoking) {
-			if (timer == null) {
-				// Start the timer for the trailing edge.
-				timer = setTimeout(timerExpired, wait);
-				return result;
-			}
-		}
```
Can you elaborate why this logic is being removed?

sergeibbb replied:
@axosoft-ramint 

This piece of code is removed because it did nothing. It didn't break thing, though. Looks like there used to be a special edging case handling that have been removed later. And this code looks like a piece of logic that just hasn't been removed but without special execution for edging case it does nothing.

axosoft-ramint commented on src/plus/launchpad/launchpadProvider.ts:
```diff
@@ -609,15 +644,52 @@ export class LaunchpadProvider implements Disposable {
 	}
 
+	mergeSearchedCategorizedItems(
+		permanent: LaunchpadCategorizedResult,
```
What do we mean by `permanent` here?

sergeibbb replied:
fixed: https://github.com/gitkraken/vscode-gitlens/pull/3684/commits/d44f3443cb11d938c01a71a65e24c20be09db999

axosoft-ramint commented on src/plus/launchpad/launchpadProvider.ts:
```diff
@@ -609,15 +644,52 @@ export class LaunchpadProvider implements Disposable {
+		}
+
+		const result = [];
+		const ids1 = new Set();
```
What is this set representing? Can we get a clearer name for it?

axosoft-ramint commented on src/plus/integrations/providers/github.ts:
```diff
@@ -317,3 +318,53 @@ export class GitHubEnterpriseIntegration extends GitHubIntegrationBase<SelfHoste
+}
+
+export function doesPullRequestSatisfyGitHubRepositoryURLIdentity(
+	pr: LaunchpadPullRequest | undefined,
```
Including Launchpad-specific types here muddles the purity and intent of the provider - instead, since this appears to be a utility fn, we can probably move it to `pullRequest.ts`, and either now or in the near future, make the parameters less GitHub-specific.

sergeibbb replied:
@axosoft-ramint 

Fixed. And it's possible to use `EnrichablePullRequest` here instead of `LaunchpadPullRequest`.

axosoft-ramint commented on src/plus/integrations/providers/github.ts:
```diff
@@ -317,3 +318,53 @@ export class GitHubEnterpriseIntegration extends GitHubIntegrationBase<SelfHoste
 	}
 }
+
+export type GitHubPullRequestURLIdentity = {
+	ownerAndRepo?: string;
+	prNumber?: string;
+};
+
+export function getPullRequestIdentityValuesFromSearch(search: string): GitHubPullRequestURLIdentity {
+	let ownerAndRepo: string | undefined = undefined;
+	let prNumber: string | undefined = undefined;
+
+	let match = search.match(/([^/]+\/[^/]+)\/pull\/(\d+)/); // with org and rep name
+	if (match != null) {
+		ownerAndRepo = match[1];
+		prNumber = match[2];
+	}
+
+	if (prNumber == null) {
+		match = search.match(/(?:\/|^)pull\/(\d+)/); // without repo name
+		if (match != null) {
+			prNumber = match[1];
+		}
+	}
+
+	if (prNumber == null) {
+		match = search.match(/(?:#|\/)(\d+)/); // any number starting with # or /
+		if (match != null) {
+			prNumber = match[1];
+		}
+	}
+
+	return { ownerAndRepo: ownerAndRepo, prNumber: prNumber };
+}
```
Same as with the function below, I suggest moving these to `pullRequest.ts` and, as soon as we start considering other providers, making them more generic/flexible.

Perhaps then the GitHub-specific regexes we are using can stay in GitHub-domain files like this one, and be imported there.

sergeibbb replied:
@axosoft-ramint 

fixed

axosoft-ramint reviewed (commented):
I didn't manage to get through all the logic in `launchpad.ts` yet, so these comments are not yet complete, but cover all the other changes. Mostly just nitpicks and questions at this point - will test more thoroughly on Monday.

axosoft-ramint commented:
Accidentally requested my own review on this PR. I removed it - feel free to re-request my review whenever it's ready to go.

sergeibbb commented:
Hi @axosoft-ramint 

I think it's ready to review.

axosoft-ramint commented on CHANGELOG.md:
```diff
@@ -34,6 +34,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/) and this p
 - Adds a new `${authorFirst}` and `${authorLast}` commit formatting tokens that can be used in inline blame, commit hovers, etc &mdash; closes [#2980](https://github.com/gitkraken/vscode-gitlens/issues/2980)
 - Adds a new _Create New Branch_ button to the _Commit Graph_ toolbar &mdash; closes [#3553](https://github.com/gitkraken/vscode-gitlens/issues/3553)
 - Adds new ability to force push from the _Commit Graph_ toolbar&mdash; closes [#3493](https://github.com/gitkraken/vscode-gitlens/issues/3493)
+- Adds new ability to search for a GitHub PR in the _Launchpad;_ closes [#3543](https://github.com/gitkraken/vscode-gitlens/issues/3543)
```
This changelog line should go in the "Unreleased" section of the changelog. If you need to add an "Added" subheading there, you can.

sergeibbb replied:
@axosoft-ramint 
My bad. I just didn't expected to meet a fresh "Unreleased" section. 

Fixed.

axosoft-ramint commented on src/plus/launchpad/launchpad.ts:
```diff
@@ -493,30 +513,51 @@ export class LaunchpadCommand extends QuickCommand<State> {
 			};
 		}
 
+		const mergeQuickpickItems = <T extends { item: { id: string } } | object>(arr: readonly T[], items: T[]) => {
```
Nitpick: let's call it `combineQuickpickItemsWithSearchResults` since 1. it's clearer and 2. we want to avoid using `merge` because it's associated elsewhere in this file with merging PR branches. And `arr` param can be `searchResults`.

Also, we discussed this elsewhere but, it would be great if the existing launchpad items can stay at the top, and the search results can be added below them. Currently they get mixed in with the results:

![search mixup](https://github.com/user-attachments/assets/375de19c-ad83-4d6d-8ed6-cedab7dd373b)

@eamodio not sure how we can get around it as it feels like we can't really control that ordering when you have input in the box. Unless you have any suggestions I don't think we should spent too much time on it as there are other priorities for this release.

sergeibbb replied:
> let's call it combineQuickpickItemsWithSearchResults

Fixed.

> Currently they get mixed in with the results

Currently VS Code does not give us a way to control it.

axosoft-ramint reviewed (commented):
This is getting close! I found one issue while testing: on the PR items that are search results, we should not have "pin" and "snooze" buttons/options, neither in this stage nor in the next stage after choosing the item (you will need to rebase this on `main` to see the "pin" and "snooze" options in the next step since that is new).

Currently, clicking "pin" or "snooze" will do effectively nothing, as we cannot show categories until the search is empty, nor should those search results persist. I suggest hiding all pin and snooze options on launchpad items that are search results.

sergeibbb commented:
@axosoft-ramint 

> on the PR items that are search results, we should not have "pin" and "snooze" buttons/options, neither in this stage nor in the next stage after choosing the item (you will need to rebase this on main to see the "pin" and "snooze" options in the next step since that is new).

1. Do you mean the state of searching? Such as removing "pin" and "snooze" from **every item.**
2. Or **only items that are "additional"** so they don't belong to the current Launchpad set of PRs?

If # 1. 

What if I want to pin an item but I don't see it in the launchpad, so I start searching it, then when it becomes clearly visible I click "pin". So, it becomes pinned.

> Currently, clicking "pin" or "snooze" will do effectively nothing, as we cannot show categories until the search is empty, nor should those search results persist

Yes. We don't see the effect immediately, but the item has been affected, so in the non-searching state of the Launchpad we'll see the result of our action.

If # 2.
Probably the main problem is that we get "involves:@me" PRs from the GitHub. So, even if I've pinned a PR, so its ID is stored on GK.dev as "pinned", I won't see it in the Launchpad, because it's outside of the set of PRs where I'm involved.

However, it would make sense if we let users pin any PR.

But now I agree, it makes sense to remove these action from "additional" items that have been found.

Thoughts?
cc @eamodio @justinrobots

axosoft-ramint reviewed (approved):
LGTM
